src/ast_helpers.hpp

```diff
@@ -0,0 +1,348 @@
+    }
+  };
+
+  // ###########################################################################
+  // Implement ordering operations for AST Nodes
+  // ###########################################################################
+
+  template <class T>
+  // Compare the objects behind pointers
+  bool PtrObjLessThanFn(const T* lhs, const T* rhs) {
+    if (lhs == nullptr) return rhs != nullptr;
+    else if (rhs == nullptr) return false;
+    else return *lhs < *rhs;
+  }
+
+  struct PtrObjLessThan {
+    template <class T>
+    // Compare the objects behind pointers
+    bool operator() (const T* lhs, const T* rhs) const {
+      return PtrObjLessThanFn<T>(lhs, rhs);
+    }
+  };
+
+  template <class T>
+  // Compare the objects and its content
+  bool ObjLessThanFn(const T& lhs, const T& rhs) {
+    return PtrObjLessThanFn(lhs.ptr(), rhs.ptr());
+  };
+
+  struct ObjLessThan {
+    template <class T>
+    // Compare the objects and its content
+    bool operator() (const T& lhs, const T& rhs) const {
+      return ObjLessThanFn<T>(lhs, rhs);
+    }
+  };
+
+  // ###########################################################################
+  // Some STL helper functions
+  // ###########################################################################
+
+  // Check if all elements are equal
+  template <class X, class Y,
+    typename XT = typename X::value_type,
+    typename YT = typename Y::value_type>
+  bool ListEquality(const X& lhs, const Y& rhs,
+    bool(*cmp)(const XT*, const YT*))
+  {
+    return lhs.size() == rhs.size() &&
+      std::equal(lhs.begin(), lhs.end(),
+        rhs.begin(), cmp);
+  }
+
+  // Return if Vector is empty
+  template <class T>
+  bool listIsEmpty(T* cnt) {
+    return cnt && cnt->empty();
+  }
+
+  // Erase items from vector that match predicate
+  template<class T, class UnaryPredicate>
+  void listEraseItemIf(T& vec, UnaryPredicate* predicate)
+  {
+    vec.erase(std::remove_if(vec.begin(), vec.end(), predicate), vec.end());
+  }
+
+  // Check that every item in `lhs` is also in `rhs`
+  // Note: this works by comparing the raw pointers
+  template <typename T>
+  bool listIsSubsetOrEqual(const T& lhs, const T& rhs) {
+    for (auto const& item : lhs) {
+      if (std::find(rhs.begin(), rhs.end(), item) == rhs.end())
+        return false;
+    }
+    return true;
+  }
+
+  // ##########################################################################
+  // Special case insensitive string matcher. We can optimize
+  // the more general compare case quite a bit by requiring
+  // consumers to obey some rules (lowercase and no space).
+  // - `literal` must only contain lower case ascii characters
+  // there is one edge case where this could give false positives
+  // test could contain a (non-ascii) char exactly 32 below literal
+  // ##########################################################################
+  inline bool equalsLiteral(const std::string& literal, const std::string test) {
```

Comment:
The second argument should be passed by reference.
The first argument should probably be a `const char *`, based on the usage.

Consider a more descriptive name, such as `equalsCaseInsensitive`.

Reply:
I would like to keep the name less general, since one should know that it is in fact not a real `equalsCaseInsensitive` method. I copied and adopted it from code we have in the parser. It does the job well enough, but only for certain input. I will change the prototype as you suggested though, I believe that it does make it more clear that this function is a little special.